internal/website/gatherexamples/gatherexamples.go

```diff
@@ -0,0 +1,271 @@
+	flag.Usage = func() {
+		out := flag.CommandLine.Output()
+		fmt.Fprintln(out, "usage: gatherexamples [options] DIR [...]")
+		fmt.Fprintln(out)
+		fmt.Fprintln(out, "Options:")
+		flag.PrintDefaults()
+	}
+	outputPath := flag.String("o", "", "path to output file (default to stdout)")
+	pattern := flag.String("pattern", "./...", "pattern to use at each directory")
+	flag.Parse()
+	if flag.NArg() == 0 {
+		flag.Usage()
+		os.Exit(2) // matches with flag package
+	}
+
+	// Load packages in each module named on the command line and find
+	// all examples.
+	allExamples := make(map[string]string)
+	for _, dir := range flag.Args() {
+		cfg := &packages.Config{
+			Mode:  gatherLoadMode,
+			Dir:   dir,
+			Tests: true,
+		}
+		pkgs, err := packages.Load(cfg, *pattern)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "gatherexamples: load %s: %v\n", dir, err)
+			os.Exit(1)
+		}
+		for exampleName, source := range gather(pkgs) {
+			allExamples[exampleName] = source
+		}
+	}
+
+	// Write all examples as a JSON object.
+	data, err := json.MarshalIndent(allExamples, "", "\t")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "gatherexamples: generate JSON: %v\n", err)
+		os.Exit(1)
+	}
+	data = append(data, '\n')
+
+	output := os.Stdout
+	if *outputPath != "" && *outputPath != "-" {
+		var err error
+		output, err = os.Create(*outputPath)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "gatherexamples: %v\n", err)
+			os.Exit(1)
+		}
+		defer func() {
+			if err := output.Close(); err != nil {
+				fmt.Fprintf(os.Stderr, "gatherexamples: write output: %v\n", err)
+				os.Exit(1)
+			}
+		}()
+	}
+	if _, err := output.Write(data); err != nil {
+		fmt.Fprintf(os.Stderr, "gatherexamples: write output: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+const gatherLoadMode packages.LoadMode = packages.NeedName |
+	packages.NeedFiles |
+	packages.NeedTypes |
+	packages.NeedSyntax |
+	packages.NeedTypesInfo
+
+func gather(pkgs []*packages.Package) map[string]string {
+	examples := make(map[string]string)
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Syntax {
+			for _, decl := range file.Decls {
+				// Determine whether this declaration is an example function.
+				fn, ok := decl.(*ast.FuncDecl)
+				if !ok {
+					continue
+				}
+				if !ok || !strings.HasPrefix(fn.Name.Name, "Example") || len(fn.Type.Params.List) > 0 || len(fn.Type.Params.List) > 0 {
+					continue
+				}
+
+				// Gather sorted list of imported packages.
+				usedPackages := make(map[string]string)
+				ast.Inspect(fn.Body, func(node ast.Node) bool {
+					id, ok := node.(*ast.Ident)
+					if !ok {
+						return true
+					}
+					refPkg, ok := pkg.TypesInfo.ObjectOf(id).(*types.PkgName)
+					if !ok {
+						return true
+					}
+					overrideName := ""
+					if id.Name != refPkg.Imported().Name() {
+						overrideName = id.Name
+					}
+					usedPackages[refPkg.Imported().Path()] = overrideName
+					return true
+				})
+
+				// Format example into string.
+				sb := new(strings.Builder)
```

Comment:
OOC: why use `new` here?

Reply:
Seemed clearer to me than `var sb strings.Builder`, since the methods all take method receivers. The compiler escape analysis makes the two equivalent anyway.